webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,309 @@
+const STORAGE_KEY = 'notebook-config';
+
+type DispatchFunction =
+  (Dispatch<{
+    key: keyof NotebookConfig,
+    value: string | number | undefined
+  }>)
+
+function reducer(
+  state: NotebookConfig,
+  action: {key: keyof NotebookConfig, value: string | number | undefined},
+): NotebookConfig {
+  return { ...state, [action.key]: action.value };
+}
+
+const useNotebookForm = (): [NotebookConfig, DispatchFunction] => {
+  const storage = useStorage(STORAGE_PATH);
+  const [ state, dispatch ] = useReducer(
+    reducer,
+    storage.getWithDefault(STORAGE_KEY, { slots: 1 }),
+  );
+
+  const storeConfig = useCallback((values: NotebookConfig) => {
+    delete values.name;
+    storage.set(STORAGE_KEY, values);
+  }, [ storage ]);
+
+  useEffect(() => {
+    storeConfig(state);
+  }, [ state, storeConfig ]);
+
+  return [ state, dispatch ];
+};
+
+interface NotebookModalProps extends ModalProps {
+  onLaunch?: () => void;
+  visible?: boolean;
+}
+
+interface FormProps {
+  fields: NotebookConfig;
+  onChange: DispatchFunction;
+}
+
+interface FullConfigProps {
+  config: RawJson;
+  onChange: (config: RawJson) => void;
+}
+
+interface ResourceInfo {
+  hasCPU: boolean;
+  hasGPU: boolean;
+  showResourceType: boolean;
+}
+
+const NotebookModal: React.FC<NotebookModalProps> = (
+  { visible = false, onLaunch, ...props }: NotebookModalProps,
+) => {
+
+  const [ showFullConfig, setShowFullConfig ] = useState(false);
+  const [ fields, dispatch ] = useNotebookForm();
+  const [ config, setConfig ] = useState<RawJson>({});
+
+  const fetchConfig = useCallback(async () => {
+    try {
+      const config = await previewNotebook(
+        fields.slots,
+        fields.template,
+        fields.name,
+        fields.pool,
+      );
+      setConfig({ config: yaml.dump(config) });
+
+    } catch {}
+  }, [ fields ]);
+
+  useEffect(() => {
+    if (showFullConfig) fetchConfig();
+  }, [ showFullConfig, fetchConfig ]);
+
+  const handleSecondary = useCallback(() => {
+    setShowFullConfig(show => !show);
+  }, [ ]);
+
+  const handleCreateEnvironment = useCallback(
+    () => {
+      if (showFullConfig) {
+        launchNotebook(config);
+      } else {
+        launchNotebook(
+          undefined,
+          fields.slots,
+          fields.template,
+          fields.name,
+          fields.pool,
+        );
+      }
+      if (onLaunch) onLaunch();
+    },
+    [ showFullConfig, onLaunch, fields, config ],
+  );
+
+  return <Modal
+    footer={<>
+      <Button onClick={handleSecondary}>{showFullConfig ? 'Edit Form' : 'Edit Full Config'}</Button>
+      <Button
+        type="primary"
+        onClick={handleCreateEnvironment}>Launch</Button>
+    </>}
+    title="Launch JupyterLab"
+    visible={visible}
+    width={540}
+    {...props}>
+    {showFullConfig ?
+      <NotebookFullConfig config={config} onChange={(newConfig) => setConfig(newConfig)} /> :
+      <NotebookForm fields={fields} onChange={dispatch} />
```

Comment:
**suggestion:** Let's leave this change for later (after this PR lands).
Looking at `NotebookModal` it's really large at this point with several large components defined within this file (`NotebookFullConfig`, `NotebookForm`, `NotebookModal`). I would recommend breaking this up and reorganizing as such:
```
src/components/NotebookModal.tsx
src/components/NotebookModal/NotebookFullConfig.tsx
src/components/NotebookModal/NotebookForm.tsx
```